Feature/bq snowflake support

emielver opened the pull request:
## Description & motivation
Add support for BigQuery (Close #7)
Add support for Snowflake (Close #8)
Bump copyright to 2022 (Close #6)

## Checklist
- [x] I have verified that these changes work locally
- [x] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)

bill-warner reviewed (commented):
Good work on this. Shaping up nicely. Left a fair few comments throughout. Mainly cleaning up redundant code plus a few tweaks.

Couple of other things to change/note:

* Update [this line](https://github.com/snowplow/dbt-snowplow-mobile/blob/b3134292b530e0eea6991f59ee838a0beb605ec9/docs/markdown/snowplow_mobile_overview.md?plain=1#L26)
* [This](https://github.com/snowplow/dbt-snowplow-mobile/blob/b3134292b530e0eea6991f59ee838a0beb605ec9/docs/markdown/snowplow_mobile_overview.md?plain=1#L540) isnt correct for screen view dedupe. Sorry I missed this in the last pass but I believe we take the first duplicate irrespective of whether there is another dupe event with the same timestamp. Reach out to me if it isnt clear why we have differing dedupe logic. 
* Update the page view table [here](https://github.com/snowplow/dbt-snowplow-mobile/blob/b3134292b530e0eea6991f59ee838a0beb605ec9/selectors.yml#L19)
* Update the custom module example to work across all warehouses. I believe the `bool_or` will fail on Snowflake and BQ. 
* I had the `snowplow_tests_screen_view_in_session_values` test fail when I ran the model 3 times from a start date of `2021-03-21` on BQ. It failed due to a repeated `screen_view_in_session_index` across different screen views within the session `cf4e0096-16c8-4be7-b614-9358d9c8fc9e`. I haven't dug into the cause of this yet but certainly worth exploring and trying to fix. I have left the table as is under my schema `dbt_will_derived` if you wanted to look into it. I suspect what has happened is the session in question spanned two runs of the model. During the second run, the initial screen view was not reprocessed.
* There is a gitkeep file in the test dir that can be removed

The failing test is my main concern here. Would be good to figure the issue out.

emielver commented on models/base/src_base.yml:
```diff
@@ -15,11 +15,15 @@ sources:
             tests:
-              - not_null
+              - not_null:
+                  config:
```
Yeah I think we could remove them and either check in the `snowplow_mobile_base_events_this_run` table or not check at all. Since we already test in that table I'll remove the tests in the source table

emielver commented on README.md:
```diff
@@ -14,11 +14,11 @@ Please refer to the [doc site][snowplow-mobile-docs] for a full breakdown of the
 ### Adapter Support
 
-The snowplow-mobile v0.1.0 package currently supports Redshift & Postgres.
+The snowplow-mobile v0.1.0 package currently supports BigQuery, Redshift, Snowflake & Postgres.
```
I'd prefer bumping all versions in the release commit so leaving this for now

emielver commented:
Thanks for the review @bill-warner. I've managed to solve/look into most of the things you were asking about. Some points:
- With regards to the readme, I'll update those version numbers once we're in the release branch
- Wrt using the `snowplow__events` variable, it looks like `dbt` treats variables that reference sources weirdly and I've come up with a workaround that hopefully you can agree with, let me know.
- I still need to test this but as far as I understand we *won't* reach conflicts when running both the mobile and web model and using the `cluster_by_fields` macros since in the dispatchers in each repo, we add the namespace we want to search (namely the package we're in), so the conflicts shouldn't exist. Nonetheless I'll double check this before asking for another go at reviewing again. 
- Thanks for the comment on the docs wrt the de-duping behaviour. As far as I can tell actually the de-duping behaviour is the same between Redshift/Postgres and Snowflake/BQ. Do you agree? If so I'll update the docs to reflect that
- Lastly wrt the bug you identified, it seems to be in the `snowplow_mobile_base_sessions_lifecycle_manifest` table. What we're seeing is the following for that session id
```
Row | event_id | session_id | screen_view_id | collector_tstamp | derived_tstamp | screen_view_id_index |
1 | e646decc-a39a-41a8-b659-eb1ca96e9829 | cf4e0096-16c8-4be7-b614-9358d9c8fc9e | eb4e1dd8-4da5-4e64-96a9-0d78a3db6576 | 2021-04-14 09:32:51.655 UTC | 2021-04-14 09:32:49.256 UTC | 1 |  
2 | de851c77-0290-494f-be19-1a61deaa7aa8 | cf4e0096-16c8-4be7-b614-9358d9c8fc9e | 41eedf0e-d45d-4a6d-a489-94ecf2a393a4 | 2021-04-14 09:33:06.876 UTC | 2021-04-14 09:33:03.325 UTC | 1 |  
3 | f1c4776d-2d8e-45e8-af48-f80f81170052 | cf4e0096-16c8-4be7-b614-9358d9c8fc9e | 7ec4865b-97e8-47dd-973e-3d359213e30d | 2021-04-14 09:33:06.876 UTC | 2021-04-14 09:33:03.062 UTC | 1 |

```
Since for the manifest we take the `min` and `max` of the `collector_tstamp`, that results in the `start_tstamp` for `session_id = 'cf4e0096-16c8-4be7-b614-9358d9c8fc9e'` to be `2021-04-14 09:32:51.655 UTC`, and the `end_tstamp = 2021-04-14 09:33:06.876 UTC`. 
As you can see, this then excludes the first `screen_view_id` from being considered. I'm not sure what the best approach is to remedy this situation, what do you think? Should we compute `min(collector_tstamp) as min_collector` and `min(derived_tstamp) as min_derived` and then store the `LOWER(min_collector, min_derived) as start_tstamp`, and the same but flipped for `end_tstamp`?

bill-warner commented:
>I still need to test this but as far as I understand we won't reach conflicts when running both the mobile and web model and using the cluster_by_fields macros since in the dispatchers in each repo, we add the namespace we want to search (namely the package we're in), so the conflicts shouldn't exist. Nonetheless I'll double check this before asking for another go at reviewing again.

Does this hold true in the situation whereby you want to change the cluster keys on both your mobile and web sessions table, but want to use differing keys across both? 

>Thanks for the comment on the docs wrt the de-duping behaviour. As far as I can tell actually the de-duping behaviour is the same between Redshift/Postgres and Snowflake/BQ. Do you agree? If so I'll update the docs to reflect that

No I don't think they are the same. For Snowflake and GBQ when deduping either the `event_id` or `screen_view_id`, we keep the first row per id ordered by `collector_tstamp` or `derived_tstamp` respectively.  

For Redshift/Postgres when deduping the `event_id`, we take the first row ordered by `collector_tstamp`. The difference here is that if there are multiple 'first' rows i.e. matching `collector_tstamp` then we discard the event altogether. Due to the federated table design of RS, any accompanying context/unstruct event tables will also have a corresponding duplicate. Because we join on `event_id` and `collector_tstamp`, the case where we have multiple 'first' rows would result in a cartesian join despite having deduped the main events table. The obvious solution to this is to also dedupe all context tables, however a) this comes at a performance hit and b) this relies upon the user following a similar practice when building custom modules involving custom contexts or events. As a result it seemed easier to simply remove these cases given there low %. For RS screen views however, the same dedupe logic as Snowflake/BQ is applied as there are no such joins. Let me know if you have any questions on that. I appreciate that is pretty dense.

>As you can see, this then excludes the first screen_view_id from being considered.

Sorry could you elaborate on this slightly more? Is it being excluded because of this filter?
https://github.com/snowplow/dbt-snowplow-mobile/blob/04994ca2e0faeb46f85d7190691172fe6b95ffb8/models/base/scratch/bigquery/snowplow_mobile_base_events_this_run.sql#L77-L80
I think this is likely the issue. For the run of the model where this error occurs the limits render as so:
```SQL
  and a.collector_tstamp >= 
    cast('2021-04-14 09:32:51.655000+00:00' as 
    timestamp
)
  
  and a.collector_tstamp <= 
    cast('2021-04-21 12:50:44.237000+00:00' as 
    timestamp
)
  
   --BQ Only
    and a.derived_tstamp >= 
    cast('2021-04-14 09:32:51.655000+00:00' as 
    timestamp
)
  
    and a.derived_tstamp <= 
    cast('2021-04-21 12:50:44.237000+00:00' as 
    timestamp
)
```

The missing problematic screen view:
![image](https://user-images.githubusercontent.com/71764352/154442893-bc9fc026-10dc-4f38-8353-f5f231e3e406.png)

I think as you are suggesting, because the limits are based on `collector_tstamp` but then we have this additional filter on `derived_tstamp` we are excluding this event. If this is indeed true, I think there are two things we can do:

a) Add some lookback period to the `lower_limit` for the `derived_tstamp` filter. For context this filter is only there for legacy pipelines which used to be partitioned by `derived_tstamp`. Then the filter on `collector_tstamp` ensure we only capture the intended events, but the filter on `derived_tstamp` ensure we prune partitions.

b) Set `snowplow__derived_tstamp_partitioned` to default to `false`. Slightly conflicted about this as it means for some legacy pipelines the model won't run out the box however it reduces the number of columns that need to be scanned for non-legacy pipelines (assuming they didn't set this value to `false` already). I believe this config is called out in the quick start guide anyway.

bill-warner commented on models/base/scratch/bigquery/snowplow_mobile_base_events_this_run.sql:
```diff
@@ -0,0 +1,99 @@
+          enabled=true,
+          col_prefix='unstruct_event_com_snowplowanalytics_mobile_screen_view_1_',
+          fields=screen_view_event_fields(),
+          relation=source('atomic','events'),
```
Sorry I thought for some reason that issue only applied to dbt utils `star` macro due to the specific relation test they have in it. Nice idea - certainly seems a like a good solution.

bill-warner replied:
This outlines the exact issue we are seeing: https://getdbt.slack.com/archives/C2JRRQDTL/p1588612231079400

emielver commented:
> Does this hold true in the situation whereby you want to change the cluster keys on both your mobile and web sessions table, but want to use differing keys across both?

Just tested this and it works. Luckily dbt already thought about this use case so we're all good on that front!

> No I don't think they are the same. For Snowflake and GBQ when deduping either the event_id or screen_view_id, we keep the first row per id ordered by collector_tstamp or derived_tstamp respectively.

Yep my bad, I was only looking at `screen_view_id` which was the same. I misunderstood your original statement and was quite confused as to why they were different, but I see now you wrote that the statement that they were different was wrong, so I've corrected that and I understand the difference in logic between `event_id`s and `screen_view_id`s

> Sorry could you elaborate on this slightly more? Is it being excluded because of this filter?
...
a) Add some lookback period to the lower_limit for the derived_tstamp filter. For context this filter is only there for legacy pipelines which used to be partitioned by derived_tstamp. Then the filter on collector_tstamp ensure we only capture the intended events, but the filter on derived_tstamp ensure we prune partitions.

I've implemented option a) and that seems to be working for the dataset, thanks for bringing it to my attention! I've also removed two tests for `not_null` for `screen_id` and `screen_name` since those were failing for the dataset. I'm not sure whether they should be implemented but it should maybe be more of a warning anyways. 

All tests are passing now on multiple runs over the dataset and with the `web` model involved as well so hopefully this should be all good to go!

bill-warner commented on models/base/scratch/bigquery/snowplow_mobile_base_events_this_run.sql:
```diff
@@ -0,0 +1,95 @@
+  inner join {{ ref('snowplow_mobile_base_sessions_this_run') }} as b
+  on {{ session_id }} = b.session_id
+
+  where {{ snowplow_utils.timestamp_diff('a.collector_tstamp', 'b.start_tstamp', 'day')}} <= {{ var("snowplow__max_session_days", 3) }}
```
Just noticed these two filters are incorrect too. I have described the issue here: https://github.com/snowplow/dbt-snowplow-web/issues/11

Very subtle difference but worth correcting I think. Seems like it applies to all 3 warehouses

bill-warner commented on models/optional_modules/app_errors/app_errors.yml:
```diff
@@ -99,8 +99,16 @@ models:
+        tests:
+          - not_null:
+              config:
+                enabled: '{{var("snowplow__enable_app_errors_module", false)}}'  
```
I think both `screen_id` and `screen_name` come from the screen context which is optional and therefore these tests could fail. We could add an extra condition that `snowplow__enable_screen_context is true`

bill-warner commented on models/optional_modules/app_errors/scratch/snowflake/snowplow_mobile_app_errors_this_run.sql:
```diff
@@ -0,0 +1,89 @@
+{{ 
+  config(
+    cluster_by = ["derived_tstamp"],
```
Up until now I have avoided clustering the `this_run` tables for Snowflake, and only clustered the derived tables. There is a cost associated to reclustering in Snowflake + clustering is best suited to very large table, which the this run tables are not really. I am not too precious about this though

bill-warner commented on macros/bool_or.sql:
```diff
@@ -0,0 +1,23 @@
+{% macro bool_or() %}
```
Not very strongly opinionated on this but it feels like having arg like `field` would be a more logical API.

```SQL
{{ bool_or(field) }}
#rather than
{{ bool_or() }}(field)
```

bill-warner commented on macros/bool_or.sql:
```diff
@@ -0,0 +1,23 @@
+
+  {{ return("LOGICAL_OR") }}
+
+{% endmacro %}
```
No newline

bill-warner commented on models/optional_modules/app_errors/scratch/app_errors_scratch.yml:
```diff
@@ -99,8 +99,16 @@ models:
       - name: screen_id
         description: '{{ doc("col_screen_id") }}'
+        tests:
+          - not_null:
```
Same here

bill-warner reviewed (commented):
Thanks for those changes - good work. Think we are almost there.

emielver commented on models/base/manifest/default/snowplow_mobile_base_sessions_lifecycle_manifest.sql:
```diff
@@ -0,0 +1,89 @@
+    and e.platform in ('{{ var("snowplow__platform")|join("','") }}') -- filters for 'mob' by default
+    and {{ is_run_with_new_events }} --don't reprocess sessions that have already been processed.
+    {% if var('snowplow__derived_tstamp_partitioned', true) and target.type == 'bigquery' | as_bool() %} -- BQ only
+      and cast(e.derived_tstamp as datetime) >= {{ dbt_utils.dateadd('hour', -1 * var('snowplow__lookback_window_hours', 6), lower_limit) }}
```
Thanks for the tip Will, I've updated the lines to be using `timestamp_add` from the `snowplow_utils` package. I've read the link and I agree it states that we shouldn't be dynamically filtering on dates, however on running the two (one with dynamically calculated dates and the other with fixed dates), there's no noticeable difference in any of the runtimes or memory usage.
Here you can see the results with dates hardcoded (between `2021-03-21` and `2021-04-21`)
![image](https://user-images.githubusercontent.com/2262032/154771422-eaabee80-3f67-4518-85f3-ca3664e3faeb.png)

versus the same but using the `timestamp_add` function:
![image](https://user-images.githubusercontent.com/2262032/154771538-4fcf04b9-9709-4649-880a-2c0a3d795695.png)

I reckon it might have to do with the `collector_tstamp` filter which is based on a harcoded timestamp, and therefore I would imagine the partitioning occurs there.

bill-warner replied:
Nice thanks for updating. Just to demonstrate the difference: 
Casting then filtering:
![image](https://user-images.githubusercontent.com/71764352/154986474-7a64e3d8-bf8e-4862-b01a-a022e9018067.png)
The double filter on `collector_tstamp` is simply because the events table requires a filter on this columns, and BQ doesn't recognise the casted column as such a filter.
No casting:
![image](https://user-images.githubusercontent.com/71764352/154986631-a3995803-7553-45c6-bf13-74b7c79d4bff.png)

Note the difference in GB scanned.

You are right though, the dynamic timestamp calc doesnt impact partition pruning:

![image](https://user-images.githubusercontent.com/71764352/154987349-335ec858-fec9-40b5-b0a6-2117b727dd4c.png)
vs.
![image](https://user-images.githubusercontent.com/71764352/154987409-80b633d9-5a4b-465e-b9d9-0c8caae38a91.png)

emielver commented:
Will thanks again for the review, I've had a look at everything and have changed stuff accordingly. The only thing I wasn't able to figure out yet was the `events` issue (https://github.com/snowplow/dbt-snowplow-mobile/pull/9#discussion_r809829797), but hopefully we can either leave it for now and open up an issue or delay the release until I'm back. Happy with either outcome, sorry I couldn't round it off today!

bill-warner commented on models/base/scratch/bigquery/snowplow_mobile_base_events_this_run.sql:
```diff
@@ -66,13 +61,13 @@ with events as (
 
-  where {{ snowplow_utils.timestamp_diff('a.collector_tstamp', 'b.start_tstamp', 'day')}} <= {{ var("snowplow__max_session_days", 3) }}
-  and {{ snowplow_utils.timestamp_diff('a.dvce_created_tstamp', 'a.dvce_sent_tstamp', 'day') }} <= {{ var("snowplow__days_late_allowed", 3) }}
+  where {{ snowplow_utils.timestamp_add('day', var("snowplow__max_session_days", 3), 'a.collector_tstamp') }} <= b.start_tstamp
```
I think this is the wrong way around and will also not allow for partition pruning. I think we can copy the logic from the web model

bill-warner commented on models/base/scratch/bigquery/snowplow_mobile_base_events_this_run.sql:
```diff
@@ -66,13 +61,13 @@ with events as (
-    and cast(a.derived_tstamp as datetime) >= {{ dbt_utils.dateadd('hour', -1 * var('snowplow__lookback_window_hours', 6), lower_limit) }}
-    and cast(a.derived_tstamp as datetime) <= {{ dbt_utils.dateadd('hour', var('snowplow__lookback_window_hours', 6), upper_limit) }}
+    and a.derived_tstamp >= {{ snowplow_utils.timestamp_add('hour', -1, lower_limit) }}
+    and a.derived_tstamp <= {{ snowplow_utils.timestamp_add('hour', 1, upper_limit) }}
```
Not we need to expand the upper limit as the `collector_tstamp` should always be greater than the `derived_tstamp`

bill-warner commented on models/base/scratch/redshift_postgres/snowplow_mobile_base_events_this_run.sql:
```diff
@@ -31,8 +31,8 @@ with events_this_run AS (
 
-  where {{ snowplow_utils.timestamp_diff('e.collector_tstamp', 'str.start_tstamp', 'day')}} <= {{ var("snowplow__max_session_days", 3) }}
-  and {{ snowplow_utils.timestamp_diff('e.dvce_created_tstamp', 'e.dvce_sent_tstamp', 'day') }} <= {{ var("snowplow__days_late_allowed", 3) }}
+  where {{ snowplow_utils.timestamp_add('day', var("snowplow__max_session_days", 3), 'e.collector_tstamp') }} <= str.start_tstamp
```
Same here. Think we need to swap this around

bill-warner commented on models/sessions/scratch/snowplow_mobile_sessions_aggs.sql:
```diff
@@ -40,7 +40,7 @@ with events as (
     MIN(e.derived_tstamp) as start_tstamp,
     MAX(e.derived_tstamp) as end_tstamp,
-    {{ bool_or() }}(e.event_name = 'application_install') as has_install
+    {{ snowplow_mobile.bool_or("e.event_name = 'application_install'") }} as has_install
```
Nice

bill-warner commented on models/base/scratch/snowflake/snowplow_mobile_base_events_this_run.sql:
```diff
@@ -103,8 +103,8 @@ with events as (
 
-  where {{ snowplow_utils.timestamp_diff('a.collector_tstamp', 'b.start_tstamp', 'day')}} <= {{ var("snowplow__max_session_days", 3) }}
-  and {{ snowplow_utils.timestamp_diff('a.dvce_created_tstamp', 'a.dvce_sent_tstamp', 'day') }} <= {{ var("snowplow__days_late_allowed", 3) }}
+  where {{ snowplow_utils.timestamp_add('day', var("snowplow__max_session_days", 3), 'a.collector_tstamp') }} <= b.start_tstamp
```
Swap here as well

bill-warner reviewed (commented):
Thanks for making those changes. They look good! I have highlighted an issue around some of the date filters. In your absence I will look to make these changes.

I still can't think of a solution to the issue around using vars to set sources. If nothing comes to mind soon, I think we just role with it as is and address this problem later. It is a pretty minor change from the users perspective.

bill-warner reviewed (approved):
In your absence I have made a final tweak to the date filters. At this point I think we are good to release. I will tidy up the commits and merge into the release branch. Good work @emielver!
